examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+        # However, don't draw it when at the selection stage and not hovering
+        # a selectable entity. Doing so would suggest you could click to select
+        # a point and proceed to the next tool stage.
+        if picked_first_point? || (@hovered_entity && @valid_hovered_entity)
+          @mouse_ip.draw(view)
+          view.tooltip = @mouse_ip.tooltip if @mouse_ip.valid?
+        end
+      end
+
+      def getExtents
+        bounds = Geom::BoundingBox.new
+        bounds.add(picked_points) unless picked_points.empty?
+        bounds
+      end
+
+      private
+
+      # Check if an entity is allowed to be selected by this tool.
+      def valid_entity?(entity)
+        # In this somewhat artificial example we only allow components to be
+        # moved. In a real life example you'd typically filter entities by
+        # other criteria.
+
+        # To replicate native Move, Rotate and Scale tools, allow all
+        # DrawingElements except Axes.
+        ### !entity.is_a?(Sketchup::Axes)
+
+        # Another typical use case is a custom tool that edits some property of
+        # a custom objects, e.g. Bezier curve control points or wall corner
+        # points. For this, you'd typically check for the entity being a group
+        # and/or component, and if it has the attribute dictionary holding your
+        # custom properties.
+        ### entity.is_a?(Sketchup::Group) && entity.attribute_dictionary("TurboWall2000")
+
+        entity.is_a?(Sketchup::ComponentInstance)
+      end
+
+      # Check if the selection as a whole is valid for this tool.
+      def valid_selection?
+        selection = Sketchup.active_model.selection
+
+        selection.size == 1 && valid_entity?(selection.first)
+      end
+
+      # Try selecting the entity under the mouse cursor, if the tool allows it.
+      def try_select_entity(view, x, y)
+        view.model.selection.clear
+
+        pickhelper = view.pick_helper(x, y)
+        @hovered_entity = pickhelper.best_picked
+        return unless @hovered_entity
+
+        @valid_hovered_entity = valid_entity?(@hovered_entity)
+        return unless @valid_hovered_entity
+
+        view.model.selection.add(@hovered_entity)
+      end
+
+      def start_move(view)
+        # HACK: create a new input point with a given position rather than
+        # copying the existing one. This fixes an issue in some cases where
+        # the points appear to get linked and the @picked_first_ip position
+        # changes when you move the mouse around, even if that should only
+        # affect @mouse_ip.
+        ### @picked_first_ip.copy!(@mouse_ip)
+        @picked_first_ip = Sketchup::InputPoint.new(@mouse_ip.position)
+
+        # Wrap model changes in an operation so it can be undone in one step.
+        # Typically we disable drawing in operations to increase performance,
+        # but here we want live updates when moving the mouse.
+        view.model.start_operation("Move", false)
```

Comment:
Worth noting that observers will not still while the operation is open. In case someone tried to use observers along with their tool while doing a long operation like this.

Reply:
I don't know if we should introduce that topic here. The reader may not know what observers are and get more confused by it. Also i think it's a quite specific case to use observer along with your custom tools.

Reply:
Fair point.